src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -2343,22 +2398,36 @@ public boolean waitForSchemaAgreement(long maxWait, TimeUnit unit, BooleanSuppli
      */
     public boolean hasMajorVersion3Nodes()
     {
-        return isUpgradingFromVersionLowerThan(CassandraVersion.CASSANDRA_4_0) || // this is quite obvious
-               // however if we discovered only nodes at current version so far (in particular only this node),
-               // but still there are nodes with unknown version, we also want to report that the cluster may have nodes at 3.x
-               upgradeInProgressPossible && !isUpgradingFromVersionLowerThan(SystemKeyspace.CURRENT_VERSION.familyLowerBound.get());
+        return isUpgradingFromVersionLowerThan(CassandraVersion.CASSANDRA_4_0);
     }
 
     /**
      * Returns {@code true} if there are nodes on version lower than the provided version
      */
     public boolean isUpgradingFromVersionLowerThan(CassandraVersion referenceVersion)
     {
-        CassandraVersion v = upgradeFromVersionMemoized.get();
-        if (CassandraVersion.NULL_VERSION.equals(v) && scheduledGossipTask == null)
-            return false;
+        return getMinVersion().compareTo(referenceVersion) < 0;
```

Comment:
this returns possibly invalid value when versions of some nodes are not known yet. getMinVersion() will return NULL_VERSION which always compares smaller than the reference version

Reply:
yes, as this method returns the lowest possible version in the cluster - if we are still during upgrade, we don't know what are the versions of nodes which are down